fsql/deser.py

```diff
@@ -109,65 +147,85 @@ def __init__(self, input_format=InputFormat.AUTO, **pdread_kwargs):
             "engine": "openpyxl",
         }
 
-    def read_single(self, partition: Partition, fs: AbstractFileSystem) -> pd.DataFrame:
-        logger.debug(f"read dataframe for partition {partition}")
-        input_format = self.detect_format(partition.url)
-        logger.debug(f"format detected for partition {input_format} <- {partition}")
+    def format_to_reader(self, input_format: InputFormat) -> Callable:
         if input_format is InputFormat.PARQUET:
-            reader = pd.read_parquet
+            return pd.read_parquet
         elif input_format is InputFormat.JSON:
-            reader = pd.read_json
+            return pd.read_json
         elif input_format is InputFormat.CSV:
-            reader = pd.read_csv
+            return pd.read_csv
         elif input_format is InputFormat.XLSX:
-            reader = pd.read_excel
+            return pd.read_excel
         elif input_format is InputFormat.AUTO:
-            raise ValueError(f"partition had format detected as auto -> invalid state. Partition: {partition}")
+            raise ValueError("partition had format detected as auto -> invalid state.")
         else:
             assert_exhaustive_enum(input_format)
 
+    def read_single(self, partition: Partition, fs: AbstractFileSystem) -> PartitionReadOutcome:
+        logger.debug(f"read dataframe for partition {partition}")
+        input_format = self.detect_format(partition.url)
+        logger.debug(f"format detected for partition {input_format} <- {partition}")
+        reader = self.format_to_reader(input_format)
+
         pdread_kwargs = {**self.pdread_default_kwargs[input_format], **self.pdread_user_kwargs}
         logger.debug(f"reader kwargs {pdread_kwargs} for partition {partition}")
-        try:
+
+        def read_dataframe(partition: Partition) -> PartitionReadOutcome:
```

Comment:
I understand, but since this method is, contextwise, internal to the encompassing one, it made more sense to me to keep it in here -- I mean, if I could declare it as a lambda, I would :)

Reply:
from this point of view, should not be  `format_to_reader` also here? I can live with a short nested functions (even lambdas), but this one has already 10 lines which just doubles the length of the method (read_single) and makes it harder to read. And regarding "internal", I agree only partially, in the future, there could be another `read_sth` method that would use the same `read_dataframe` function (or do you think there is not such possible case?).

Reply:
well I factored `format_to_reader` to a standalone function primarily because the function's complexity breached allowed limit :innocent: Otherwise, I would keep it inside / turn it into an extension of the enum itself...

and the _only_ reason for the read_dataframe existing is the need to retry because of the FileNotFound exception, otherwise it would not be needed...

I understand that those two paragraphs are not consistent with each other -- I shall give a thought to whether I can get rid of the format_to_reader method somehow

Reply:
not really happy with this in any way... There seems to lurk some similarity in the way how the read_dataframe/read_dict are structured invoked from within the respective read_single; but I prefer the code repetition to premature abstraction here... At some point later, it may make sense to, instead of read_single, ask for reader function, exception list, and object+partitions->object functions, and plug those three generically, but I'm not ready for that just yet. I'll leave a design note there

I did rename both to `_read_internal`, and I changed the `format_to_reader` to `@caseclass`, but thats just cosmetics/whim